sdk/messaging/azservicebus/internal/retrier.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package internal
+
+import (
+	"context"
+	"time"
+
+	"github.com/jpillora/backoff"
+)
+
+// TODO: we should discuss what a common policy would be.
+var DefaultRetryPolicy Retrier = &BackoffRetrier{
+	Backoff: backoff.Backoff{
+		Factor: 1,
+		Min:    time.Second * 5,
+	},
+	MaxRetries: 5,
+}
+
+// A retrier that allows you to do a basic for loop and get backoff
+// and retry limits. See `Try` for more details on how to use it.
+type Retrier interface {
+	// Copies the retrier. Retriers are stateful and must be copied
+	// before starting a set of retries.
+	Copy() Retrier
+
+	// Returns true if the retries were exhausted.
+	Exhausted() bool
+
+	// Try marks an attempt to call (first call to Try() does not sleep).
+	// Will return false if the `ctx` is cancelled or if we exhaust our retries.
+	//
+	//    rp := RetryPolicy{Backoff:defaultBackoffPolicy, MaxRetries:5}
+	//
+	//    for rp.Try(ctx) {
+	//       <your code>
+	//    }
+	//
+	//    if rp.Cancelled() || rp.Exhausted() {
+	//       // no more retries needed
+	//    }
+	//
+	Try(ctx context.Context) bool
+}
+
+// Encapsulates a backoff policy, which allows you to configure the amount of
+// time in between retries as well as the maximum retries allowed (via MaxRetries)
+// NOTE: this should be copied by the caller as it is stateful.
+type BackoffRetrier struct {
+	Backoff    backoff.Backoff
```

Comment:
I'd prefer if we didn't use anything from `backoff` as an exchange type.

Reply:
Does it change things if this is only in the `internal` package?

Reply:
No, because we don't control the definition of `backoff.Backoff` which might change in the future.

Reply:
Made the entire `BackoffRetrier` an internal type instead.